### DISCUSSION TITLE
choco install : An item with the same key has already been added.

### DISCUSSION THREAD

**Discussion by t794104** [chocolatey/choco#2617] [Q&A]
We have created a package that for some reason when we try too install it gives us the error : "An item with the same key has already been added.".

No leads in the logging; so no clue why.

`choco install isw-swh-1 --version 0.1.0-jb0036 --pre -v -d
Chocolatey v0.11.2
Chocolatey is running on Windows v 10.0.14393.0
Attempting to delete file "C:/ProgramData/chocolatey/choco.exe.old".
Attempting to delete file "C:\ProgramData\chocolatey\choco.exe.old".
Command line: "C:\ProgramData\chocolatey\choco.exe" install isw-swh-1 --version 0.1.0-jb0036 --pre -v -d
Received arguments: install isw-swh-1 --version 0.1.0-jb0036 --pre -v -d
RemovePendingPackagesTask is now ready and waiting for PreRunMessage.
Sending message 'PreRunMessage' out if there are subscribers...
[Pending] Removing all pending packages that should not be considered installed...
Performing validation checks.
Global Configuration Validation Checks:
 - Package Exit Code / Exit On Reboot = Checked
System State Validation Checks:
 Reboot Requirement Checks:
 - Pending Computer Rename = Checked
 - Pending Component Based Servicing = Checked
 - Pending Windows Auto Update = Checked
 - Pending File Rename Operations = Ignored
 - Pending Windows Package Installer = Checked
 - Pending Windows Package Installer SysWow64 = Checked
The source 'https://jfrog.internal.vodafoneziggo.com/artifactory/api/nuget/isw-choco' evaluated to a 'normal' source type

NOTE: Hiding sensitive configuration data! Please double and triple
 check to be sure no sensitive data is shown, especially if copying
 output to a gist for review.
Configuration: CommandName='install'|
CacheLocation='C:\Users\srvISWd\AppData\Local\Temp\chocolatey'|
ContainsLegacyPackageInstalls='True'|
CommandExecutionTimeoutSeconds='2700'|WebRequestTimeoutSeconds='30'|
Sources='https://jfrog.internal.vodafoneziggo.com/artifactory/api/nuget/isw-choco'|

SourceType='normal'|Debug='True'|Verbose='True'|Trace='False'|
Force='False'|Noop='False'|HelpRequested='False'|
UnsuccessfulParsing='False'|RegularOutput='True'|QuietOutput='False'|
PromptForConfirmation='False'|AcceptLicense='False'|
AllowUnofficialBuild='False'|Input='isw-swh-1'|Version='0.1.0-jb0036'|
AllVersions='False'|SkipPackageInstallProvider='False'|
PackageNames='isw-swh-1'|Prerelease='True'|ForceX86='False'|
OverrideArguments='False'|NotSilent='False'|
ApplyPackageParametersToDependencies='False'|
ApplyInstallArgumentsToDependencies='False'|IgnoreDependencies='False'|
AllowMultipleVersions='False'|AllowDowngrade='False'|
ForceDependencies='False'|Information.PlatformType='Windows'|
Information.PlatformVersion='10.0.14393.0'|
Information.PlatformName='Windows Server 2016'|
Information.ChocolateyVersion='0.11.2.0'|
Information.ChocolateyProductVersion='0.11.2'|
Information.FullName='choco, Version=0.11.2.0, Culture=neutral, PublicKeyToken=79d02ea9cad655eb'|

Information.Is64BitOperatingSystem='True'|
Information.Is64BitProcess='True'|Information.IsInteractive='True'|
Information.UserName='srvISWd'|Information.UserDomainName='RESOURCE'|
Information.IsUserAdministrator='True'|
Information.IsUserSystemAccount='False'|
Information.IsUserRemoteDesktop='False'|
Information.IsUserRemote='True'|
Information.IsProcessElevated='True'|
Information.IsLicensedVersion='False'|Information.LicenseType='Foss'|
Information.CurrentDirectory='C:\Users\srvISWd\Downloads'|
Features.AutoUninstaller='True'|Features.ChecksumFiles='True'|
Features.AllowEmptyChecksums='False'|
Features.AllowEmptyChecksumsSecure='True'|
Features.FailOnAutoUninstaller='False'|
Features.FailOnStandardError='False'|Features.UsePowerShellHost='True'|
Features.LogEnvironmentValues='False'|Features.LogWithoutColor='False'|
Features.VirusCheck='False'|
Features.FailOnInvalidOrMissingLicense='False'|
Features.IgnoreInvalidOptionsSwitches='True'|
Features.UsePackageExitCodes='True'|
Features.UseEnhancedExitCodes='False'|
Features.UseFipsCompliantChecksums='False'|
Features.ShowNonElevatedWarnings='True'|
Features.ShowDownloadProgress='True'|
Features.StopOnFirstPackageFailure='False'|
Features.UseRememberedArgumentsForUpgrades='True'|
Features.IgnoreUnfoundPackagesOnUpgradeOutdated='False'|
Features.SkipPackageUpgradesWhenNotInstalled='False'|
Features.RemovePackageInformationOnUninstall='False'|
Features.ExitOnRebootDetected='False'|
Features.LogValidationResultsOnWarnings='True'|
Features.UsePackageRepositoryOptimizations='True'|
Features.ScriptsCheckLastExitCode='False'|
ListCommand.LocalOnly='False'|
ListCommand.IdOnly='False'|ListCommand.IncludeRegistryPrograms='False'|
ListCommand.PageSize='25'|ListCommand.Exact='False'|
ListCommand.ByIdOnly='False'|ListCommand.ByTagOnly='False'|
ListCommand.IdStartsWith='False'|ListCommand.OrderByPopularity='False'|
ListCommand.ApprovedOnly='False'|
ListCommand.DownloadCacheAvailable='False'|
ListCommand.NotBroken='False'|
ListCommand.IncludeVersionOverrides='False'|
UpgradeCommand.FailOnUnfound='False'|
UpgradeCommand.FailOnNotInstalled='False'|
UpgradeCommand.NotifyOnlyAvailableUpgrades='False'|
UpgradeCommand.ExcludePrerelease='False'|
NewCommand.AutomaticPackage='False'|
NewCommand.UseOriginalTemplate='False'|SourceCommand.Command='unknown'|
SourceCommand.Priority='0'|SourceCommand.BypassProxy='False'|
SourceCommand.AllowSelfService='False'|
SourceCommand.VisibleToAdminsOnly='False'|
FeatureCommand.Command='unknown'|ConfigCommand.Command='unknown'|
ApiKeyCommand.Remove='False'|PinCommand.Command='unknown'|
OutdatedCommand.IgnorePinned='False'|
ExportCommand.IncludeVersionNumbers='False'|Proxy.BypassOnLocal='True'|
_ Chocolatey:ChocolateyInstallCommand - Normal Run Mode _
Installing the following packages:
isw-swh-1
By installing, you accept licenses for the packages.
Attempting to gather credentials for 'https://jfrog.internal.vodafoneziggo.com/artifactory/api/nuget/isw-choco'
Using saved credentials
Progress: Downloading isw-swh-1 0.1.0-jb0036... 100%
isw-swh-1 not installed. An error occurred during installation:
 An item with the same key has already been added.
isw-swh-1 package files install failed with exit code 1. Performing other installation steps.
No package information as package is null.
No package information to save as package is null.
Sending message 'HandlePackageResultCompletedMessage' out if there are subscribers...
The install of isw-swh-1 was NOT successful.
isw-swh-1 not installed. An error occurred during installation:
 An item with the same key has already been added.

Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Failures
 - isw-swh-1 (exited 1) - isw-swh-1 not installed. An error occurred during installation:
 An item with the same key has already been added.
Sending message 'PostRunMessage' out if there are subscribers...
Exiting with 1
`

**Comment by t794104** [comment]
Added info; we found out that it was a file included in the package. The extension had capitals and after changing them too lower the error was resolved.

Question: if we include 2 files with the same extension but different capitals ('.seproject' and '.SeProject') does that break Choco ?

**Reply by TheCakeIsNaOH** [reply]
> @t794104 Question: if we include 2 files with the same extension but different capitals ('.seproject' and '.SeProject') does that break Choco ?

So like `file.seproject` and `file.SeProject` file in the same folder? Or something like `fileA.seproject` and `fileB.SeProject`?

**Reply by t794104** [reply]
I'm wondering about both cases ? but for use it was the later one. Differnt filename same extention (exept caps).

**Reply by gep13** [reply]
Are you able to provide the complete package that isn't working (i.e. the packaging scripts that you are using chocolateyInstall.ps1, *.nuspec, etc)? Or can you create a small reproduction of the problem that you are seeing?  That would be the best way forward in terms of figuring out what is going on.

**Reply by t794104** [reply]
we fixed it in this package and moved on. 

I'll try and create a example setup that produces the same error and upload that here

**Reply by gep13** [reply]
That would be great, thank you!

**Comment by TheCakeIsNaOH** [comment]
> but for use it was the later one. Differnt filename same extention (exept caps).

That should not be a problem. I've installed packages with this, and not seen an issue. If it is actually the problem that caused the error you saw, it is something that should be fixed.

> I'm wondering about both cases ?

So for the first case, where there are two files in the same folder, but with different casing:
- On Windows, it should not be able to be setup in the first place, due to the non case sensitive file system. So you couldn't put to files with the same name (including extension) but different cases.
- On Linux, it should not be able to be packed to create the package, it should fail with `This partUri is already contained in the package` when running `choco pack`
